File: 300_Longest_Increasing_Subsequence/lengthOfLIS.py
class Solution(object):
    def lengthOfLIS(self, nums):
        if not nums or len(nums) == 1: return len(nums)
        f = [1] * len(nums)

        for i in xrange(1, len(nums)):
            for j in xrange(i):
                if nums[i] > nums[j]:
                    f[i] = max(f[i], f[j] + 1)

        #print f
        return max(f)

sol = Solution()
#nums = [1,3,6,7,9,4,10,5,6]
nums = [3, 100, 2, 104, 3, 4, 5, 6]
#nums = [3, 100, 2, 104]
print nums
print sol.lengthOfLIS(nums)

File: 300_Longest_Increasing_Subsequence/lengthOfLIS.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
    public:
        int lengthOfLIS(vector<int>& nums) {
            int N = nums.size();
            if(N == 0) return 0;
            vector<int> LIS;
            LIS.push_back(nums[0]);

            for(int i = 1; i < N; i++) {
                if(nums[i] > LIS.back())
                    LIS.push_back(nums[i]);
                else {
                    int begin = 0, end = LIS.size() - 1;
                    while(begin <= end) {
                        int mid = begin + (end - begin) / 2;
                        if(nums[i] == LIS[mid]) {
                            begin = mid;
                            break;
                        }
                        if(nums[i] > LIS[mid])
                            begin = mid + 1;
                        else
                            end = mid - 1;
                    }
                    LIS[begin] = nums[i];
                }
            }
            return LIS.size();
        }
};

int main() {
    return 0;
}
